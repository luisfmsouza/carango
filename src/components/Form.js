import React from "react";
import styled from "styled-components";

import { ReactComponent as IconRight } from "../assets/icons/arrow-right.svg";

import colors from "../theme/colors";

const Form = styled.form``;

const Column = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  grid-column-gap: 20px;

  @media (min-width: 700px) {
    grid-template-columns: 1fr 1fr;
  }
`;

const Group = styled.div`
  margin-top: 30px;
  position: relative;
`;

const Label = styled.label`
  color: ${colors.grey20};
  display: block;
  font-family: LeasePlan, Arial, Helvetica, sans-serif;
  font-weight: 600;
  font-size: 16px;
  line-height: 1.5;

  ${({ hasError }) => hasError && `color: ${colors.red30}`}
`;

const InputContainer = styled.div`
  background-color: ${colors.white};
  border: 1px solid ${colors.grey15};
  border-radius: 2px;
  position: relative;
  width: 100%;

  ${({ hasError }) => hasError && `background-color: ${colors.red10}`}
  ${({ hasError }) => hasError && `border-color: ${colors.red30}`}

  &::after {
    ${({ hasError }) =>
      hasError &&
      `
        background-color: ${colors.red20};
        border-radius: 11px;
        box-sizing: border-box;
        color: ${colors.red30};
        content: "!";
        display: block;
        height: 22px;
        padding: 3px 0;
        position: absolute;
        right: 13px;
        text-align: center;
        top: 15px;
        width: 22px;
      `}
  }
`;

const Input = styled.input`
  background-color: transparent;
  border: 0;
  box-sizing: border-box;
  font-family: LeasePlan, Arial, Helvetica, sans-serif;
  font-size: 16px;
  line-height: 1.5;
  padding: 15px 50px 15px 15px;
  width: 100%;
`;

const ErrorMessage = styled.span`
  color: ${colors.red30};
  display: block;
  font-family: LeasePlan, Arial, Helvetica, sans-serif;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.5;
  margin-top: 8px;
`;

const CheckboxLabel = styled.label`
  cursor: pointer;
  margin-left: 8px;
  vertical-align: middle;

  ${({ hasError }) => hasError && `color: ${colors.red30}`}
`;

const CheckboxInput = styled.input`
  cursor: pointer;
  height: 22px;
  margin: 0;
  opacity: 0;
  order: 1;
  position: absolute;
  vertical-align: middle;
  width: 22px;
  z-index: 1;

  &:checked ~ span::after {
    color: ${colors.grey20};
  }
`;

const CheckboxShape = styled.span`
  vertical-align: middle;

  &::before {
    background-color: ${colors.white};
    border: 1px solid ${colors.grey15};
    border-radius: 2px;
    content: "";
    display: inline-block;
    height: 20px;
    width: 20px;

    ${({ hasError }) => hasError && `background-color: ${colors.red10}`}
    ${({ hasError }) => hasError && `border-color: ${colors.red30}`}
  }

  &::after {
    color: ${colors.white};
    content: "";
    border-bottom: 2px solid;
    border-left: 2px solid;
    height: 6px;
    left: 5px;
    position: absolute;
    top: 5px;
    transform: rotate(-45deg);
    width: 10px;

    ${({ hasError }) => hasError && `color: ${colors.red10}`}
  }
`;

const Button = styled.button`
  background: linear-gradient(to right, ${colors.orange10}, ${colors.orange20});
  border-width: 0px;
  border-radius: 50px;
  cursor: pointer;
  color: ${colors.white};
  font-family: LeasePlan, Arial, Helvetica, sans-serif;
  font-size: 14px;
  font-weight: 600;
  letter-spacing: 1px;
  margin-top: 30px;
  padding: 18px 25px 18px 35px;
  text-transform: uppercase;
`;

const ButtonText = styled.span`
  margin-right: 15px;
  vertical-align: middle;
`;

const StyledIconRight = styled(IconRight)`
  fill: ${colors.white};
  vertical-align: middle;
  width: 15px;
`;

export default ({ hasError }) => (
  <Form>
    <Column>
      <Group>
        <Label htmlFor="form--first-name" hasError={hasError}>
          First name
        </Label>
        <InputContainer hasError={hasError}>
          <Input
            id="form--first-name"
            placeholder="Enter first name"
            type="text"
            hasError={hasError}
          />
        </InputContainer>
        {hasError && <ErrorMessage>Invalid first name</ErrorMessage>}
      </Group>

      <Group>
        <Label htmlFor="form--surname" hasError={hasError}>
          Surname
        </Label>
        <InputContainer hasError={hasError}>
          <Input
            id="form--surname"
            placeholder="Enter surname"
            type="text"
            hasError={hasError}
          />
        </InputContainer>
        {hasError && <ErrorMessage>Invalid surname</ErrorMessage>}
      </Group>
    </Column>

    <Group>
      <Label htmlFor="form--email" hasError={hasError}>
        Email address
      </Label>
      <InputContainer hasError={hasError}>
        <Input
          id="form--email"
          placeholder="Enter your email address"
          type="email"
          hasError={hasError}
        />
      </InputContainer>
      {hasError && <ErrorMessage>Invalid email address</ErrorMessage>}
    </Group>

    <Group>
      <Label htmlFor="form--phone" hasError={hasError}>
        Phone number
      </Label>
      <InputContainer hasError={hasError}>
        <Input
          id="form--phone"
          placeholder="Enter your phone number"
          type="tel"
          hasError={hasError}
        />
      </InputContainer>
      {hasError && <ErrorMessage>Invalid phone number</ErrorMessage>}
    </Group>

    <Group>
      <Label htmlFor="form--voucher" hasError={hasError}>
        Voucher code
      </Label>
      <InputContainer hasError={hasError}>
        <Input
          id="form--voucher"
          placeholder="Enter your voucher code"
          type="number"
          hasError={hasError}
        />
      </InputContainer>
      {hasError && <ErrorMessage>Invalid voucher code</ErrorMessage>}
    </Group>

    <Group>
      <CheckboxInput id="form--policy" type="checkbox" />
      <CheckboxShape hasError={hasError} />
      <CheckboxLabel htmlFor="form--policy" hasError={hasError}>
        Accept Privacy Settlement
      </CheckboxLabel>
    </Group>

    <Button type="submit">
      <ButtonText>Get Offer</ButtonText>
      <StyledIconRight />
    </Button>
  </Form>
);
