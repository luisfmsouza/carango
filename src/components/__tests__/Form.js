import React from "react";
import renderer from "react-test-renderer";

import Form from "../Form";

describe.only("Form", () => {
  it("renders the Form component", () => {
    const actual = renderer.create(<Form />).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it("renders the Form component when has error", () => {
    const actual = renderer.create(<Form hasError={true} />).toJSON();

    expect(actual).toMatchSnapshot();
  });
});
