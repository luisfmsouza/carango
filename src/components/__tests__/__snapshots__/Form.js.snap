// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders the Form component 1`] = `
<form
  className="sc-bdVaJa bDWFJH"
>
  <div
    className="sc-bwzfXH dckAAz"
  >
    <div
      className="sc-htpNat eRDnou"
    >
      <label
        className="sc-bxivhb kgdeGf"
        htmlFor="form--first-name"
      >
        First name
      </label>
      <div
        className="sc-ifAKCX bcdHbI"
      >
        <input
          className="sc-EHOje btOswH"
          id="form--first-name"
          placeholder="Enter first name"
          type="text"
        />
      </div>
    </div>
    <div
      className="sc-htpNat eRDnou"
    >
      <label
        className="sc-bxivhb kgdeGf"
        htmlFor="form--surname"
      >
        Surname
      </label>
      <div
        className="sc-ifAKCX bcdHbI"
      >
        <input
          className="sc-EHOje btOswH"
          id="form--surname"
          placeholder="Enter surname"
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb kgdeGf"
      htmlFor="form--email"
    >
      Email address
    </label>
    <div
      className="sc-ifAKCX bcdHbI"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--email"
        placeholder="Enter your email address"
        type="email"
      />
    </div>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb kgdeGf"
      htmlFor="form--phone"
    >
      Phone number
    </label>
    <div
      className="sc-ifAKCX bcdHbI"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--phone"
        placeholder="Enter your phone number"
        type="tel"
      />
    </div>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb kgdeGf"
      htmlFor="form--voucher"
    >
      Voucher code
    </label>
    <div
      className="sc-ifAKCX bcdHbI"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--voucher"
        placeholder="Enter your voucher code"
        type="number"
      />
    </div>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <input
      className="sc-htoDjs jtEFYT"
      id="form--policy"
      type="checkbox"
    />
    <span
      className="sc-dnqmqq dBKiXt"
    />
    <label
      className="sc-gzVnrw krUXAY"
      htmlFor="form--policy"
    >
      Accept Privacy Settlement
    </label>
  </div>
  <button
    className="sc-iwsKbI dWNEbR"
    type="submit"
  >
    <span
      className="sc-gZMcBi dKOCRF"
    >
      Get Offer
    </span>
    <svg
      className="sc-gqjmRU fPOneR"
    >
      arrow-right.svg
    </svg>
  </button>
</form>
`;

exports[`Form renders the Form component when has error 1`] = `
<form
  className="sc-bdVaJa bDWFJH"
>
  <div
    className="sc-bwzfXH dckAAz"
  >
    <div
      className="sc-htpNat eRDnou"
    >
      <label
        className="sc-bxivhb jjexRw"
        htmlFor="form--first-name"
      >
        First name
      </label>
      <div
        className="sc-ifAKCX hAdtJO"
      >
        <input
          className="sc-EHOje btOswH"
          id="form--first-name"
          placeholder="Enter first name"
          type="text"
        />
      </div>
      <span
        className="sc-bZQynM hzFBlF"
      >
        Invalid first name
      </span>
    </div>
    <div
      className="sc-htpNat eRDnou"
    >
      <label
        className="sc-bxivhb jjexRw"
        htmlFor="form--surname"
      >
        Surname
      </label>
      <div
        className="sc-ifAKCX hAdtJO"
      >
        <input
          className="sc-EHOje btOswH"
          id="form--surname"
          placeholder="Enter surname"
          type="text"
        />
      </div>
      <span
        className="sc-bZQynM hzFBlF"
      >
        Invalid surname
      </span>
    </div>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb jjexRw"
      htmlFor="form--email"
    >
      Email address
    </label>
    <div
      className="sc-ifAKCX hAdtJO"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--email"
        placeholder="Enter your email address"
        type="email"
      />
    </div>
    <span
      className="sc-bZQynM hzFBlF"
    >
      Invalid email address
    </span>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb jjexRw"
      htmlFor="form--phone"
    >
      Phone number
    </label>
    <div
      className="sc-ifAKCX hAdtJO"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--phone"
        placeholder="Enter your phone number"
        type="tel"
      />
    </div>
    <span
      className="sc-bZQynM hzFBlF"
    >
      Invalid phone number
    </span>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <label
      className="sc-bxivhb jjexRw"
      htmlFor="form--voucher"
    >
      Voucher code
    </label>
    <div
      className="sc-ifAKCX hAdtJO"
    >
      <input
        className="sc-EHOje btOswH"
        id="form--voucher"
        placeholder="Enter your voucher code"
        type="number"
      />
    </div>
    <span
      className="sc-bZQynM hzFBlF"
    >
      Invalid voucher code
    </span>
  </div>
  <div
    className="sc-htpNat eRDnou"
  >
    <input
      className="sc-htoDjs jtEFYT"
      id="form--policy"
      type="checkbox"
    />
    <span
      className="sc-dnqmqq eXgceN"
    />
    <label
      className="sc-gzVnrw gpBFas"
      htmlFor="form--policy"
    >
      Accept Privacy Settlement
    </label>
  </div>
  <button
    className="sc-iwsKbI dWNEbR"
    type="submit"
  >
    <span
      className="sc-gZMcBi dKOCRF"
    >
      Get Offer
    </span>
    <svg
      className="sc-gqjmRU fPOneR"
    >
      arrow-right.svg
    </svg>
  </button>
</form>
`;
